src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/WhyCommand/WhyCommandUtility/DependencyGraphFinder.cs

```diff
@@ -0,0 +1,306 @@
+
+            // hashset tracking every package node that we've traversed
+            var visited = new HashSet<string>();
+            // dictionary tracking all package nodes that have been added to the graph, mapped to their DependencyNode objects
+            var dependencyNodes = new Dictionary<string, DependencyNode>();
+            // dictionary mapping all packageIds to their resolved version
+            Dictionary<string, string> versions = GetAllResolvedVersions(packageLibraries);
+
+            foreach (var topLevelReference in topLevelReferences)
+            {
+                // use depth-first search to find dependency paths from the top-level package to the target package
+                DependencyNode? topLevelNode = FindDependencyPath(topLevelReference, packageLibraries, visited, dependencyNodes, versions, targetPackage);
+
+                if (topLevelNode != null)
+                {
+                    dependencyGraph ??= [];
+                    dependencyGraph.Add(topLevelNode);
+                }
+            }
+
+            return dependencyGraph;
+        }
+
+        /// <summary>
+        /// Traverses the dependency graph for a given top-level package, looking for a path to the target package.
+        /// </summary>
+        /// <param name="topLevelPackage">Top-level package.</param>
+        /// <param name="packageLibraries">All package libraries for a given framework.</param>
+        /// <param name="visited">HashSet tracking every package node that we've traversed.</param>
+        /// <param name="dependencyNodes">Dictionary tracking all packageIds that were added to the graph, mapped to their DependencyNode objects.</param>
+        /// <param name="versions">Dictionary mapping packageIds to their resolved versions.</param>
+        /// <param name="targetPackage">The package we want the dependency paths for.</param>
+        /// <returns>
+        /// The top-level package node in the dependency graph (if a path was found), or null (if no path was found).
+        /// </returns>
+        private static DependencyNode? FindDependencyPath(
+            string topLevelPackage,
+            IList<LockFileTargetLibrary> packageLibraries,
+            HashSet<string> visited,
+            Dictionary<string, DependencyNode> dependencyNodes,
+            Dictionary<string, string> versions,
+            string targetPackage)
+        {
+            var stack = new Stack<StackDependencyData>();
+            stack.Push(new StackDependencyData(topLevelPackage, null));
+
+            while (stack.Count > 0)
+            {
+                var currentPackageData = stack.Pop();
+                var currentPackageId = currentPackageData.Id;
+
+                // if we reach the target node, or if we've already traversed this node and found dependency paths, add it to the graph
+                if (currentPackageId == targetPackage
+                    || dependencyNodes.ContainsKey(currentPackageId))
```

Comment:
Dictionary looks ups as well. 

We need to make sure the comparer is the right one.

Reply:
Fixed this where we first declare the HashSet